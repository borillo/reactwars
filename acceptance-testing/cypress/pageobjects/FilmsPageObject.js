class FilmsPageObject {
  constructor(wrapper) {
    this.wrapper = wrapper;
  }

  open() {
    this.wrapper.server();
    this.wrapper.route("/api/films", "fixture:films");

    this.wrapper.visit("/films");
  }

  title() {
    return this.wrapper.get(".films-panel h1").first();
  }

  films() {
    return this.wrapper.get(".film");
  }

  filmsTitles() {
    return this.wrapper.get(".film .title");
  }

  filmsEpisodes() {
    return this.wrapper.get(".film .episode");
  }

  numVotes(filmName) {
    return this.wrapper
      .contains(".film", filmName)
      .find(".votes")
      .invoke("text")
      .then(text => {
        if (!text) return 0;
        return parseInt(text, 10);
      });
  }

  voteFilm(filmName) {
    this.wrapper
      .contains(filmName)
      .parentsUntil(".film")
      .parent()
      .find(".vote")
      .click();
  }

  registerFilm(film) {
    return this.wrapper
      .get(".new-film-title")
      .type(film.title)
      .get(".new-film-director")
      .type(film.director)
      .get(".new-film-year")
      .type(film.year)
      .get(".new-film-poster")
      .type(film.poster)
      .get(".save-new-film")
      .click();
  }
}

export default FilmsPageObject;
