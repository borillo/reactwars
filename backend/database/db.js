const films = [
  {
    episode: 1,
    title: "The Phantom Menace",
    year: 1999
  },
  {
    episode: 2,
    title: "Attack of the Clones",
    year: 2002
  },
  {
    episode: 3,
    title: "Revenge of the Sith",
    year: 2005
  },
  {
    episode: 4,
    title: "A New Hope",
    year: 1977
  },
  {
    episode: 5,
    title: "The Empire Strikes Back",
    year: 1980
  },
  {
    episode: 6,
    title: "Return of the Jedi",
    year: 1983
  },
  {
    episode: 7,
    title: "The Force Awakens",
    year: 2015
  },
  {
    episode: 8,
    title: "The Last Jedi",
    year: 2017
  }
];

export default {
  films: films,
  titles: films.map(film => film.title),
  episodes: films.map(film => film.episode),
  numberOfFilms: films.length
};
