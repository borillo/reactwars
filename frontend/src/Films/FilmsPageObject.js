export default class FilmsPageObject {
  constructor({ getByTestId, container, debug }) {
    this.getByTestId = getByTestId;
    this.container = container;
    this.debug = debug;
  }

  retrieveDom() {
    return this.container;
  }

  obtainFilms() {
    return this.getByTestId("films").querySelectorAll(".film");
  }

  obtainFilmsTitles() {
    const nodeTitles = this.container.querySelectorAll(".title");

    return [...nodeTitles].map(node => node.textContent);
  }

  obtainFilmsEpisodes() {
    const nodeEpisodes = this.container.querySelectorAll(".episode");

    return [...nodeEpisodes].map(node => node.textContent);
  }
}
