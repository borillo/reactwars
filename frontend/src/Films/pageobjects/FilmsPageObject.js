export default class FilmsPageObject {
  constructor({ getByTestId, container }) {
    this.getByTestId = getByTestId;
    this.container = container;
  }

  retrieveDom() {
    return this.container;
  }
  
  obtainFilms() {
    return this.getByTestId("films").children;
  }

  obtainFilmsTitles() {
    return []; /*this.wrapper
      .update()
      .find(".film .title")
      .map(film => film.text());*/
  }

  obtainFilmsEpisodes() {
    return []; /*this.wrapper
      .update()
      .find(".film .episode")
      .map(film => film.text());*/
  }
}
