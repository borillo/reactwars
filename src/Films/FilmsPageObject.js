export default class FilmsPageObject {
  constructor(wrapper) {
    this.wrapper = wrapper;
  }
 
  obtainFilms() {
    return this.wrapper.update().find('.film');
  }
  
  obtainFilmsTitles() {
    return this.wrapper.update().find('.film .title').map((film) => film.text());
  }

  obtainFilmsEpisodes() {
    return this.wrapper.update().find('.film .episode').map((film) => film.text());
  }
}
