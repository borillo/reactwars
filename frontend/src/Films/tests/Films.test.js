import componentDom from "./dom/init";
import FilmsPageObject from "../FilmsPageObject";

import db from "../../../../backend/database/db";
const FILM_TITLES = db.titles;
const NUMBER_OF_FILMS = db.numberOfFilms;

describe("Films", () => {
  let page;

  beforeEach(() => {
    page = new FilmsPageObject(componentDom);
  });

  test("should be listed", async () => {
    let filmList = page.obtainFilms();

    expect(filmList).toHaveLength(NUMBER_OF_FILMS);
  });

  test("should show name for each shown episode", async () => {
    const filmsTitles = page.obtainFilmsTitles();

    expect(filmsTitles).toEqual(FILM_TITLES);
  });

  test("should be ordered by episode number", async () => {
    let episodes = page.obtainFilmsEpisodes();
    
    episodes.forEach((episode, index) => {
      expect(episode).toEqual(`Episode ${index + 1}`);
    });
  });

  test("should match snapshot", async () => {
    expect(page.retrieveDom()).toMatchSnapshot();
  });
});
