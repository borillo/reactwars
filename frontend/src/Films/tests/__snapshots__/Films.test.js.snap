// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Films should match snapshot 1`] = `
<div>
  <div
    class="films-panel"
  >
    <h1>
      Film list
    </h1>
    <div
      class="films"
      data-testid="films"
    >
      <div
        class="films-list"
      >
        <div
          class="film"
        >
          <div
            class="poster"
          >
            <img
              alt="poster"
              src="https://upload.wikimedia.org/wikipedia/en/4/40/Star_Wars_Phantom_Menace_poster.jpg"
            />
          </div>
          <div
            class="detail"
          >
            <div
              class="title"
            >
              <a
                href="/films/1"
              >
                The Phantom Menace
              </a>
            </div>
            <div
              class="episode"
            >
              Episode 
              1
            </div>
            <div>
              <button
                class="vote"
              >
                <svg
                  aria-hidden="true"
                  data-prefix="fas"
                  data-icon="thumbs-up"
                  class="svg-inline--fa fa-thumbs-up fa-w-16 "
                  role="img"
                  xmlns="http://www.w3.org/2000/svg"
                  viewBox="0 0 512 512"
                >
                  <path
                    fill="currentColor"
                    d="M104 224H24c-13.255 0-24 10.745-24 24v240c0 13.255 10.745 24 24 24h80c13.255 0 24-10.745 24-24V248c0-13.255-10.745-24-24-24zM64 472c-13.255 0-24-10.745-24-24s10.745-24 24-24 24 10.745 24 24-10.745 24-24 24zM384 81.452c0 42.416-25.97 66.208-33.277 94.548h101.723c33.397 0 59.397 27.746 59.553 58.098.084 17.938-7.546 37.249-19.439 49.197l-.11.11c9.836 23.337 8.237 56.037-9.308 79.469 8.681 25.895-.069 57.704-16.382 74.757 4.298 17.598 2.244 32.575-6.148 44.632C440.202 511.587 389.616 512 346.839 512l-2.845-.001c-48.287-.017-87.806-17.598-119.56-31.725-15.957-7.099-36.821-15.887-52.651-16.178-6.54-.12-11.783-5.457-11.783-11.998v-213.77c0-3.2 1.282-6.271 3.558-8.521 39.614-39.144 56.648-80.587 89.117-113.111 14.804-14.832 20.188-37.236 25.393-58.902C282.515 39.293 291.817 0 312 0c24 0 72 8 72 81.452z"
                  />
                </svg>
              </button>
              <span
                class="votes"
              />
              vote for me!!!
            </div>
          </div>
        </div>
        <div
          class="film"
        >
          <div
            class="poster"
          >
            <img
              alt="poster"
              src="https://upload.wikimedia.org/wikipedia/en/3/32/Star_Wars_-_Episode_II_Attack_of_the_Clones_%28movie_poster%29.jpg"
            />
          </div>
          <div
            class="detail"
          >
            <div
              class="title"
            >
              <a
                href="/films/2"
              >
                Attack of the Clones
              </a>
            </div>
            <div
              class="episode"
            >
              Episode 
              2
            </div>
            <div>
              <button
                class="vote"
              >
                <svg
                  aria-hidden="true"
                  data-prefix="fas"
                  data-icon="thumbs-up"
                  class="svg-inline--fa fa-thumbs-up fa-w-16 "
                  role="img"
                  xmlns="http://www.w3.org/2000/svg"
                  viewBox="0 0 512 512"
                >
                  <path
                    fill="currentColor"
                    d="M104 224H24c-13.255 0-24 10.745-24 24v240c0 13.255 10.745 24 24 24h80c13.255 0 24-10.745 24-24V248c0-13.255-10.745-24-24-24zM64 472c-13.255 0-24-10.745-24-24s10.745-24 24-24 24 10.745 24 24-10.745 24-24 24zM384 81.452c0 42.416-25.97 66.208-33.277 94.548h101.723c33.397 0 59.397 27.746 59.553 58.098.084 17.938-7.546 37.249-19.439 49.197l-.11.11c9.836 23.337 8.237 56.037-9.308 79.469 8.681 25.895-.069 57.704-16.382 74.757 4.298 17.598 2.244 32.575-6.148 44.632C440.202 511.587 389.616 512 346.839 512l-2.845-.001c-48.287-.017-87.806-17.598-119.56-31.725-15.957-7.099-36.821-15.887-52.651-16.178-6.54-.12-11.783-5.457-11.783-11.998v-213.77c0-3.2 1.282-6.271 3.558-8.521 39.614-39.144 56.648-80.587 89.117-113.111 14.804-14.832 20.188-37.236 25.393-58.902C282.515 39.293 291.817 0 312 0c24 0 72 8 72 81.452z"
                  />
                </svg>
              </button>
              <span
                class="votes"
              />
              vote for me!!!
            </div>
          </div>
        </div>
        <div
          class="film"
        >
          <div
            class="poster"
          >
            <img
              alt="poster"
              src="https://upload.wikimedia.org/wikipedia/en/9/93/Star_Wars_Episode_III_Revenge_of_the_Sith_poster.jpg"
            />
          </div>
          <div
            class="detail"
          >
            <div
              class="title"
            >
              <a
                href="/films/3"
              >
                Revenge of the Sith
              </a>
            </div>
            <div
              class="episode"
            >
              Episode 
              3
            </div>
            <div>
              <button
                class="vote"
              >
                <svg
                  aria-hidden="true"
                  data-prefix="fas"
                  data-icon="thumbs-up"
                  class="svg-inline--fa fa-thumbs-up fa-w-16 "
                  role="img"
                  xmlns="http://www.w3.org/2000/svg"
                  viewBox="0 0 512 512"
                >
                  <path
                    fill="currentColor"
                    d="M104 224H24c-13.255 0-24 10.745-24 24v240c0 13.255 10.745 24 24 24h80c13.255 0 24-10.745 24-24V248c0-13.255-10.745-24-24-24zM64 472c-13.255 0-24-10.745-24-24s10.745-24 24-24 24 10.745 24 24-10.745 24-24 24zM384 81.452c0 42.416-25.97 66.208-33.277 94.548h101.723c33.397 0 59.397 27.746 59.553 58.098.084 17.938-7.546 37.249-19.439 49.197l-.11.11c9.836 23.337 8.237 56.037-9.308 79.469 8.681 25.895-.069 57.704-16.382 74.757 4.298 17.598 2.244 32.575-6.148 44.632C440.202 511.587 389.616 512 346.839 512l-2.845-.001c-48.287-.017-87.806-17.598-119.56-31.725-15.957-7.099-36.821-15.887-52.651-16.178-6.54-.12-11.783-5.457-11.783-11.998v-213.77c0-3.2 1.282-6.271 3.558-8.521 39.614-39.144 56.648-80.587 89.117-113.111 14.804-14.832 20.188-37.236 25.393-58.902C282.515 39.293 291.817 0 312 0c24 0 72 8 72 81.452z"
                  />
                </svg>
              </button>
              <span
                class="votes"
              />
              vote for me!!!
            </div>
          </div>
        </div>
        <div
          class="film"
        >
          <div
            class="poster"
          >
            <img
              alt="poster"
              src="https://upload.wikimedia.org/wikipedia/en/8/87/StarWarsMoviePoster1977.jpg"
            />
          </div>
          <div
            class="detail"
          >
            <div
              class="title"
            >
              <a
                href="/films/4"
              >
                A New Hope
              </a>
            </div>
            <div
              class="episode"
            >
              Episode 
              4
            </div>
            <div>
              <button
                class="vote"
              >
                <svg
                  aria-hidden="true"
                  data-prefix="fas"
                  data-icon="thumbs-up"
                  class="svg-inline--fa fa-thumbs-up fa-w-16 "
                  role="img"
                  xmlns="http://www.w3.org/2000/svg"
                  viewBox="0 0 512 512"
                >
                  <path
                    fill="currentColor"
                    d="M104 224H24c-13.255 0-24 10.745-24 24v240c0 13.255 10.745 24 24 24h80c13.255 0 24-10.745 24-24V248c0-13.255-10.745-24-24-24zM64 472c-13.255 0-24-10.745-24-24s10.745-24 24-24 24 10.745 24 24-10.745 24-24 24zM384 81.452c0 42.416-25.97 66.208-33.277 94.548h101.723c33.397 0 59.397 27.746 59.553 58.098.084 17.938-7.546 37.249-19.439 49.197l-.11.11c9.836 23.337 8.237 56.037-9.308 79.469 8.681 25.895-.069 57.704-16.382 74.757 4.298 17.598 2.244 32.575-6.148 44.632C440.202 511.587 389.616 512 346.839 512l-2.845-.001c-48.287-.017-87.806-17.598-119.56-31.725-15.957-7.099-36.821-15.887-52.651-16.178-6.54-.12-11.783-5.457-11.783-11.998v-213.77c0-3.2 1.282-6.271 3.558-8.521 39.614-39.144 56.648-80.587 89.117-113.111 14.804-14.832 20.188-37.236 25.393-58.902C282.515 39.293 291.817 0 312 0c24 0 72 8 72 81.452z"
                  />
                </svg>
              </button>
              <span
                class="votes"
              />
              vote for me!!!
            </div>
          </div>
        </div>
        <div
          class="film"
        >
          <div
            class="poster"
          >
            <img
              alt="poster"
              src="https://upload.wikimedia.org/wikipedia/en/3/3c/SW_-_Empire_Strikes_Back.jpg"
            />
          </div>
          <div
            class="detail"
          >
            <div
              class="title"
            >
              <a
                href="/films/5"
              >
                The Empire Strikes Back
              </a>
            </div>
            <div
              class="episode"
            >
              Episode 
              5
            </div>
            <div>
              <button
                class="vote"
              >
                <svg
                  aria-hidden="true"
                  data-prefix="fas"
                  data-icon="thumbs-up"
                  class="svg-inline--fa fa-thumbs-up fa-w-16 "
                  role="img"
                  xmlns="http://www.w3.org/2000/svg"
                  viewBox="0 0 512 512"
                >
                  <path
                    fill="currentColor"
                    d="M104 224H24c-13.255 0-24 10.745-24 24v240c0 13.255 10.745 24 24 24h80c13.255 0 24-10.745 24-24V248c0-13.255-10.745-24-24-24zM64 472c-13.255 0-24-10.745-24-24s10.745-24 24-24 24 10.745 24 24-10.745 24-24 24zM384 81.452c0 42.416-25.97 66.208-33.277 94.548h101.723c33.397 0 59.397 27.746 59.553 58.098.084 17.938-7.546 37.249-19.439 49.197l-.11.11c9.836 23.337 8.237 56.037-9.308 79.469 8.681 25.895-.069 57.704-16.382 74.757 4.298 17.598 2.244 32.575-6.148 44.632C440.202 511.587 389.616 512 346.839 512l-2.845-.001c-48.287-.017-87.806-17.598-119.56-31.725-15.957-7.099-36.821-15.887-52.651-16.178-6.54-.12-11.783-5.457-11.783-11.998v-213.77c0-3.2 1.282-6.271 3.558-8.521 39.614-39.144 56.648-80.587 89.117-113.111 14.804-14.832 20.188-37.236 25.393-58.902C282.515 39.293 291.817 0 312 0c24 0 72 8 72 81.452z"
                  />
                </svg>
              </button>
              <span
                class="votes"
              />
              vote for me!!!
            </div>
          </div>
        </div>
        <div
          class="film"
        >
          <div
            class="poster"
          >
            <img
              alt="poster"
              src="https://upload.wikimedia.org/wikipedia/en/b/b2/ReturnOfTheJediPoster1983.jpg"
            />
          </div>
          <div
            class="detail"
          >
            <div
              class="title"
            >
              <a
                href="/films/6"
              >
                Return of the Jedi
              </a>
            </div>
            <div
              class="episode"
            >
              Episode 
              6
            </div>
            <div>
              <button
                class="vote"
              >
                <svg
                  aria-hidden="true"
                  data-prefix="fas"
                  data-icon="thumbs-up"
                  class="svg-inline--fa fa-thumbs-up fa-w-16 "
                  role="img"
                  xmlns="http://www.w3.org/2000/svg"
                  viewBox="0 0 512 512"
                >
                  <path
                    fill="currentColor"
                    d="M104 224H24c-13.255 0-24 10.745-24 24v240c0 13.255 10.745 24 24 24h80c13.255 0 24-10.745 24-24V248c0-13.255-10.745-24-24-24zM64 472c-13.255 0-24-10.745-24-24s10.745-24 24-24 24 10.745 24 24-10.745 24-24 24zM384 81.452c0 42.416-25.97 66.208-33.277 94.548h101.723c33.397 0 59.397 27.746 59.553 58.098.084 17.938-7.546 37.249-19.439 49.197l-.11.11c9.836 23.337 8.237 56.037-9.308 79.469 8.681 25.895-.069 57.704-16.382 74.757 4.298 17.598 2.244 32.575-6.148 44.632C440.202 511.587 389.616 512 346.839 512l-2.845-.001c-48.287-.017-87.806-17.598-119.56-31.725-15.957-7.099-36.821-15.887-52.651-16.178-6.54-.12-11.783-5.457-11.783-11.998v-213.77c0-3.2 1.282-6.271 3.558-8.521 39.614-39.144 56.648-80.587 89.117-113.111 14.804-14.832 20.188-37.236 25.393-58.902C282.515 39.293 291.817 0 312 0c24 0 72 8 72 81.452z"
                  />
                </svg>
              </button>
              <span
                class="votes"
              />
              vote for me!!!
            </div>
          </div>
        </div>
        <div
          class="film"
        >
          <div
            class="poster"
          >
            <img
              alt="poster"
              src="https://upload.wikimedia.org/wikipedia/en/a/a2/Star_Wars_The_Force_Awakens_Theatrical_Poster.jpg"
            />
          </div>
          <div
            class="detail"
          >
            <div
              class="title"
            >
              <a
                href="/films/7"
              >
                The Force Awakens
              </a>
            </div>
            <div
              class="episode"
            >
              Episode 
              7
            </div>
            <div>
              <button
                class="vote"
              >
                <svg
                  aria-hidden="true"
                  data-prefix="fas"
                  data-icon="thumbs-up"
                  class="svg-inline--fa fa-thumbs-up fa-w-16 "
                  role="img"
                  xmlns="http://www.w3.org/2000/svg"
                  viewBox="0 0 512 512"
                >
                  <path
                    fill="currentColor"
                    d="M104 224H24c-13.255 0-24 10.745-24 24v240c0 13.255 10.745 24 24 24h80c13.255 0 24-10.745 24-24V248c0-13.255-10.745-24-24-24zM64 472c-13.255 0-24-10.745-24-24s10.745-24 24-24 24 10.745 24 24-10.745 24-24 24zM384 81.452c0 42.416-25.97 66.208-33.277 94.548h101.723c33.397 0 59.397 27.746 59.553 58.098.084 17.938-7.546 37.249-19.439 49.197l-.11.11c9.836 23.337 8.237 56.037-9.308 79.469 8.681 25.895-.069 57.704-16.382 74.757 4.298 17.598 2.244 32.575-6.148 44.632C440.202 511.587 389.616 512 346.839 512l-2.845-.001c-48.287-.017-87.806-17.598-119.56-31.725-15.957-7.099-36.821-15.887-52.651-16.178-6.54-.12-11.783-5.457-11.783-11.998v-213.77c0-3.2 1.282-6.271 3.558-8.521 39.614-39.144 56.648-80.587 89.117-113.111 14.804-14.832 20.188-37.236 25.393-58.902C282.515 39.293 291.817 0 312 0c24 0 72 8 72 81.452z"
                  />
                </svg>
              </button>
              <span
                class="votes"
              />
              vote for me!!!
            </div>
          </div>
        </div>
        <div
          class="film"
        >
          <div
            class="poster"
          >
            <img
              alt="poster"
              src="https://upload.wikimedia.org/wikipedia/en/7/7f/Star_Wars_The_Last_Jedi.jpg"
            />
          </div>
          <div
            class="detail"
          >
            <div
              class="title"
            >
              <a
                href="/films/8"
              >
                The Last Jedi
              </a>
            </div>
            <div
              class="episode"
            >
              Episode 
              8
            </div>
            <div>
              <button
                class="vote"
              >
                <svg
                  aria-hidden="true"
                  data-prefix="fas"
                  data-icon="thumbs-up"
                  class="svg-inline--fa fa-thumbs-up fa-w-16 "
                  role="img"
                  xmlns="http://www.w3.org/2000/svg"
                  viewBox="0 0 512 512"
                >
                  <path
                    fill="currentColor"
                    d="M104 224H24c-13.255 0-24 10.745-24 24v240c0 13.255 10.745 24 24 24h80c13.255 0 24-10.745 24-24V248c0-13.255-10.745-24-24-24zM64 472c-13.255 0-24-10.745-24-24s10.745-24 24-24 24 10.745 24 24-10.745 24-24 24zM384 81.452c0 42.416-25.97 66.208-33.277 94.548h101.723c33.397 0 59.397 27.746 59.553 58.098.084 17.938-7.546 37.249-19.439 49.197l-.11.11c9.836 23.337 8.237 56.037-9.308 79.469 8.681 25.895-.069 57.704-16.382 74.757 4.298 17.598 2.244 32.575-6.148 44.632C440.202 511.587 389.616 512 346.839 512l-2.845-.001c-48.287-.017-87.806-17.598-119.56-31.725-15.957-7.099-36.821-15.887-52.651-16.178-6.54-.12-11.783-5.457-11.783-11.998v-213.77c0-3.2 1.282-6.271 3.558-8.521 39.614-39.144 56.648-80.587 89.117-113.111 14.804-14.832 20.188-37.236 25.393-58.902C282.515 39.293 291.817 0 312 0c24 0 72 8 72 81.452z"
                  />
                </svg>
              </button>
              <span
                class="votes"
              />
              vote for me!!!
            </div>
          </div>
        </div>
      </div>
      <div
        class="films-film-detail"
      />
    </div>
  </div>
</div>
`;
